airbyte-cdk/python/airbyte_cdk/sources/connector_state_manager.py

```diff
@@ -14,41 +27,117 @@ class ConnectorStateManager:
+        :param value: A stream state mapping that is being updated for a stream
+        :return:
+        """
+        stream_descriptor = HashableStreamDescriptor(name=stream_name, namespace=namespace)
+        self.per_stream_states[stream_descriptor] = AirbyteStateBlob.parse_obj(value)
+
+    @classmethod
+    def _extract_from_state_message(
+        cls, state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]], stream_instance_map: Mapping[str, Stream]
+    ) -> Tuple[Optional[AirbyteStateBlob], MutableMapping[HashableStreamDescriptor, Optional[AirbyteStateBlob]]]:
+        """
+        Takes an incoming list of state messages or the legacy state format and extracts state attributes according to type
+        which can then be assigned to the new state manager being instantiated
+        :param state: The incoming state input
+        :return: A tuple of shared state, per stream state, and legacy state assembled from the incoming state list
+        """
+        if state is None:
+            return None, {}
+
+        # Incoming pure legacy object format
+        if isinstance(state, dict):
+            streams = cls._create_stream_from_state_object(state, stream_instance_map)
+            return None, streams
+
+        if not isinstance(state, List):
             raise ValueError("Input state should come in the form of list of Airbyte state messages or a mapping of states")
 
-    def get_stream_state(self, namespace: str, stream_name: str) -> AirbyteStateBlob:
-        # todo implement in upcoming PRs
-        pass
+        # When processing incoming state in source.read_state(), legacy state gets deserialized into List[AirbyteStateMessage]
+        # which can be translated into independent per-stream state values
+        if cls._is_migrated_legacy_state(state):
+            streams = cls._create_stream_from_state_object(state[0].data, stream_instance_map)
+            return None, streams
+
+        if cls._is_global_state(state):
+            global_state = state[0].global_
+            shared_state = copy.deepcopy(global_state.shared_state, {})
+            streams = {
+                HashableStreamDescriptor(
+                    name=per_state.stream_descriptor.name, namespace=per_state.stream_descriptor.namespace
+                ): per_state.stream_state
+                for per_state in global_state.stream_states
+            }
+            return shared_state, streams
 
-    def get_legacy_state(self) -> MutableMapping[str, Any]:
+        # Assuming all prior conditions were not met this is a per-stream list of states
+        streams = {
+            HashableStreamDescriptor(
+                name=per_state.stream.stream_descriptor.name, namespace=per_state.stream.stream_descriptor.namespace
+            ): per_state.stream.stream_state
+            for per_state in state
+            if per_state.type == AirbyteStateType.STREAM and hasattr(per_state, "stream")
+        }
```

Comment:
Actually I think that it makes sense because it is the default for it to be detected last. We explicitly look for specific characteristics like type == GLOBAL to identify the message format and anything that doesn't fall into those buckets will get defaulted to per-stream whenever applicable.